### PR TITLE
Various tweaks

### PR DESCRIPTION
**Pull request by lierdakil**
The main thing I decided to do this for is 5b3f13b (note if/when this is merged, atom-ide-base types need to be updated too). The reason is that purifier filters out `file:///` urls from anchors by default, which is a non-starter for one of my projects. If you think making it user-configurable is too much, another option is to set `ALLOW_UNKNOWN_PROTOCOLS: true` unconditionally. However, adding an option is more backwards-compatible and flexible, so I've chosen to go this route instead.

The other big thing, 22dbff4, is something I had to figure out with makrdown-preview-plus. Long story short, the method used for highlighting code in atom/markdown-preview (which was evidently reused here) is pretty broken. I'm using a more low-level approach that at least guarantees that all input is consumed, unlike the unbelievably hacky "make an editor and rip its guts out" approach taken previously. Both methods rely pretty heavily on undocumented Atom internals. My method at least relies on a relatively stable internal API and not on TextEditorElement implementation detail.

I've also fixed a couple compiler warnings while at it.

Note: I couldn't run tests (apparently my Atom install decided dev mode is too much of a luxury today for some reason), so if tests fail, I won't know until CI finishes.

### CODE REVIEW

**Comment by lierdakil** [comment]
Linter says: 
> Checking formatting...
[warn] modules/highlighter.d.ts
[warn] modules/highlighter.js
[warn] modules/main.d.ts
[warn] modules/main.js
[warn] modules/renderer.d.ts
[warn] modules/renderer.js
[warn] modules/utils/render.d.ts
[warn] modules/utils/render.js
[warn] modules/utils/utils.d.ts
[warn] modules/utils/utils.js
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
 ERROR  Command failed with exit code 1.

Seriously? Pretty sure if I commit those it will complain I committed those. Or do we commit generated files here? They are gitignored...

**Comment by aminya** [comment]
I have forgotten to add `modules` to .prttierignore

File: src/highlighter.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+export async function highlightTreeSitter(sourceCode: string, scopeName: string) {
+  const yielder = await eventLoopYielder(100, 5000)
+  const buf = new TextBuffer()
+  try {
+    const grammar = atom.grammars.grammarForId(scopeName)
+    const lm = atom.grammars.languageModeForGrammarAndBuffer(grammar, buf)
+    buf.setLanguageMode(lm)
+    buf.setText(sourceCode)
+    const end = buf.getEndPosition()
+    if (lm.startTokenizing) lm.startTokenizing()
```
Comment (aminya) [inline src/highlighter.ts:35]: The recent versions of eslint-config-atomic will nag about not using braces, but it is fine for now 😄

Comment (lierdakil) [reply src/highlighter.ts:35]: I personally use the rule "if it's in one line, no braces, if multiline, always braces". Works pretty well in terms of readability and doesn't create giant towers of `{}` for simple checks.

File: src/highlighter.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    const iter = lm.buildHighlightIterator()
+    if (iter.getOpenScopeIds && iter.getCloseScopeIds) {
+      let pos = { row: 0, column: 0 }
+      iter.seek(pos)
+      const res = []
+      while (pos.row < end.row || (pos.row === end.row && pos.column <= end.column)) {
+        const open = iter.getOpenScopeIds().map((x) => lm.classNameForScopeId(x))
+        const close = iter.getCloseScopeIds().map((x) => lm.classNameForScopeId(x))
+        res.push(...close.map((_) => `</span>`))
+        res.push(...open.map((x) => `<span class="${x}">`))
```
Comment (aminya) [inline src/highlighter.ts:46]: Are these two the same?
```suggestion
        res.push(
          /** open */ ...iter.getOpenScopeIds().map((x) => `<span class="${lm.classNameForScopeId(x)}">`),
          /** close */ ...iter.getCloseScopeIds().map((_) => "</span>")
        )
```

File: src/highlighter.ts (hunk 1 above)
Comment (aminya) [reply src/highlighter.ts:35]: The issue with one-line if is that you can't debug it, and the sentence is kinda hidden from the eyes. But that's fine, the tools will update it later.
https://youtu.be/GPP64opjy_Y?t=1068

Comment (lierdakil) [reply src/highlighter.ts:35]: > you can't debug it

Conditional breakpoints to the rescue. But I think stepping debuggers are overrated anyway. As for "hidden", well, no, it's really not, it's there in plain sight, as any other statement. Towers of indented blocks really force you to scan code left to right which doesn't help readability. But this is off-topic. If you want to discuss this further, I suggest we move this to Discord.

File: src/highlighter.ts (hunk 2 above)
Comment (lierdakil) [reply src/highlighter.ts:46]: We first must push `close`, then `open`, due to how iterator is structured. Other than that, probably the same in terms of behaviour. In terms of performance, I would expect V8 should fuse map invocations, but it's fickle, so maybe not.

Comment (aminya) [reply src/highlighter.ts:46]: So, the following works for sure
```suggestion
        // Note: don't inline! the iterators should be calculated in order
        const close = iter.getCloseScopeIds().map(() => "</span>")
        const open = iter.getOpenScopeIds().map((x) => `<span class="${lm.classNameForScopeId(x)}">`)
        res.push(...close, ...open)
```

Comment (aminya) [reply src/highlighter.ts:46]: or is it just pushing that should happen in order? If that is the only requirement, then the first suggestion is fine because push always pushes from the start.

Comment (lierdakil) [reply src/highlighter.ts:46]: No, no, the order of `get*ScopeIds()` invocations doesn't matter, only the actual order those are pushed to `res`. Besides, in JS arguments are evaluated left-to-right by spec.

Comment (aminya) [reply src/highlighter.ts:46]: Right, so the first suggestion works.

Comment (lierdakil) [reply src/highlighter.ts]: You've edited it after I started replying :shrug:

File: src/highlighter.ts (hunk 1 above)
Comment (aminya) [reply src/highlighter.ts:35]: I don't like the braces either, but I appreciate a new line after a condition. However, Prettier doesn't allow putting the sentences on a separate line. Either way, that's fine for now. 

> Discord
It more suits [r/programminghumor](https://www.reddit.com/r/programminghumor/) 😄

File: src/highlighter.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        disp.dispose()
+        resolve(undefined)
+      })
+    } else {
+      resolve(undefined) // null language mode
+    }
+  })
+}
+
+async function eventLoopYielder(delayMs: number, maxTimeMs: number) {
```
Comment (aminya) [inline src/highlighter.ts:82]: What does this do?

Comment (lierdakil) [reply src/highlighter.ts:82]: Yields the event loop every `delayMs` ms or so, and kills the highlighter after `maxTimeMs` ms or so (at which point it will return partial result). This is done mostly to avoid intermittent UI freezes, since all the work is being done in the main thread. And no, it's impossible to move it to another thread.

This is mostly relevant for long-ish documents, but it doesn't really hurt here either, except maybe there's a little overhead on the first yield (on the first line). We can drop that if you want.

Comment (aminya) [reply src/highlighter.ts:82]: I don't quite understand how it prevents freezing. Does this `await yielder` allow JavaScript's event loop to calculate other things? If so, I think we can use an actual generator function to call `yield`. Otherwise, if only a timer is needed, then we can use a simple timer to check if this is working?

Comment (lierdakil) [reply src/highlighter.ts:82]: `await` in general suspends the computation and yields the event loop. The point here is to suspend the computation until `setImmediate` fires (which will get queued and fire only after other queued events). You could also rewrite this directly using continuation-passing style with `setImmediate`, but cps in js looks somewhat ugly and can be memory-heavy if you're not careful.

Comment (lierdakil) [reply src/highlighter.ts:82]: Okay, sorry, probably bad explanation, I'm a little preoccupied right now so words are extra hard.

TL;DR: using await here is equivalent to passing a continuation callback to `setImmediate`. `await`ing on a synchronous computation wouldn't work.

Comment (aminya) [reply src/highlighter.ts:82]: I made a PR to your PR making the code reusable:
https://github.com/lierdakil/atom-ide-markdown-service/pull/1

**Comment by lierdakil** [comment]
@aminya ping. I really want 5b3f13b merged sooner rather than later.

**Comment by aminya** [comment]
I haven't tested this in action. Does it work?

File: src/renderer.ts [hunk 4]
```diff
@@ -7,71 +6,33 @@ import marked from "marked"
...
  * renders markdown to safe HTML asynchronously
  * @param markdownText the markdown text to render
  * @param scopeName scope name used for highlighting the code
  * @return the html string containing the result
  */
-function internalRender(markdownText: string, scopeName: string = "text.plain"): Promise<string> {
+function internalRender(
+  markdownText: string,
+  scopeName: string = "text.plain",
+  purifyConfig?: PurifyConfig
```
Comment (aminya) [inline src/renderer.ts:26]: Could you rename this to `DOMPurifyConfig` for more clarity? Please also add the description for it to the docstring.

File: src/renderer.ts [hunk 5]
```diff
@@ -7,71 +6,33 @@ import marked from "marked"
...
-    el.remove()
-  }
-}
+import { highlightTreeSitter } from "./highlighter"
 
 marked.setOptions({
   breaks: true,
 })
 
+export type PurifyConfig = Omit<DOMPurify.Config, "RETURN_DOM" | "RETURN_DOM_FRAGMENT" | "RETURN_TRUSTED_TYPE">
```
Comment (aminya) [inline src/renderer.ts:15]: Why should we omit these?

Comment (lierdakil) [reply src/renderer.ts:15]: Those change argument and return types of the purifier function. Since we're pretty much expecting to work with strings here, I don't think the complications are worth it. Hence, those are omitted.

Comment (aminya) [reply src/renderer.ts:15]: Is there any complication? The output is returned straight. Just remove the return type and let TypeScript infer it.

Comment (lierdakil) [reply src/renderer.ts:15]: 1) This will considerably complicate types in atom-ide-base, which would have to reimplement the logic behind these options. IMO the increased complexity isn't worth it, since these options do not affect functionality, and only slightly affect ergonomics.
2) The documentation insists that some combinations of these options have, to quote them, "foot-gun potential".

**Comment by lierdakil** [comment]
> I haven't tested this in action. Does it work?

I don't generally post nontrivial pull requests without testing that in my code. Now, I didn't do any _comprehensive_ testing, but I've been running the modified version since before I created the pull request and didn't encounter any issues.

**Comment by lierdakil** [comment]
Side note, is there a reason to have `render` and `renderInternal` where the former is basically an alias for the latter?

**Comment by lierdakil** [comment]
I've merged `render` and `internalRender` in eb52e8a. Also, the docstring for `render` was lying: `@param grammar the default grammar used in code sections that have no specific grammar set` was wrong, it is actually `@param scopeName scope name used for highlighting the code`, i.e. it will be used unconditionally (incidentally that's why `lang` parameter is unused, which was explicitly marked as such in bc3d790).

You could either treat it as a documentation issue or a bug depending on disposition. I suggest avoid changing the behaviour, and update the documentation instead (eb52e8a does that by virtue of getting rid of duplication)

File: src/renderer.ts (hunk 5 above)
Comment (aminya) [reply src/renderer.ts]: OK. That is fine for now.

**Comment by aminya** [comment]
@lierdakil Could reword this commit to be a "feat"?

https://github.com/atom-community/atom-ide-markdown-service/pull/43/commits/22dbff4e620ad94d8cedb8fdf6a700d51f010911

If you are away from the keyboard, I can do this as well.
